src/lib/arch/riscv/Instruction.cc

```diff
@@ -0,0 +1,172 @@
+    auto reg = registers[i];
+    if (reg == Instruction::ZERO_REGISTER) {
+      // Any zero-register references should be marked as ready, and
+      //  the corresponding operand value zeroed
+      operands[i] = RegisterValue(0, 8);
+      operandsPending--;
+    }
+    sourceRegisters[i] = reg;
+  }
+}
+void Instruction::setDestinationRegisters(
+    const std::vector<Register>& registers) {
+  assert(registers.size() <= MAX_DESTINATION_REGISTERS &&
+         "Exceeded maximum destination registers for a RISCV instruction");
+  destinationRegisterCount = registers.size();
+  std::copy(registers.begin(), registers.end(), destinationRegisters.begin());
+}
+
+const span<Register> Instruction::getOperandRegisters() const {
+  return {const_cast<Register*>(sourceRegisters.data()), sourceRegisterCount};
+}
+const span<Register> Instruction::getDestinationRegisters() const {
+  return {const_cast<Register*>(destinationRegisters.data()),
+          destinationRegisterCount};
+}
+bool Instruction::isOperandReady(int index) const {
+  return static_cast<bool>(operands[index]);
+}
+
+void Instruction::renameSource(uint8_t i, Register renamed) {
+  sourceRegisters[i] = renamed;
+}
+void Instruction::renameDestination(uint8_t i, Register renamed) {
+  destinationRegisters[i] = renamed;
+}
+
+void Instruction::supplyOperand(uint8_t i, const RegisterValue& value) {
+  assert(!canExecute() &&
+         "Attempted to provide an operand to a ready-to-execute instruction");
+  assert(value.size() > 0 &&
+         "Attempted to provide an uninitialised RegisterValue");
+
+  operands[i] = value;
+  operandsPending--;
+}
+
+void Instruction::supplyData(uint64_t address, const RegisterValue& data) {
+  for (size_t i = 0; i < memoryAddresses.size(); i++) {
+    if (memoryAddresses[i].address == address && !memoryData[i]) {
+      if (!data) {
+        // Raise exception for failed read
+        // TODO: Move this logic to caller and distinguish between different
+        // memory faults (e.g. bus error, page fault, seg fault)
+        exception_ = InstructionException::DataAbort;
+        exceptionEncountered_ = true;
+        memoryData[i] = RegisterValue(0, memoryAddresses[i].size);
+      } else {
+        memoryData[i] = data;
+      }
+      dataPending_--;
+      return;
+    }
+  }
+}
+
+span<const RegisterValue> Instruction::getData() const {
+  return {memoryData.data(), memoryData.size()};
+}
+
+bool Instruction::canExecute() const { return (operandsPending == 0); }
+
+const span<RegisterValue> Instruction::getResults() const {
+  return {const_cast<RegisterValue*>(results.data()), destinationRegisterCount};
+}
+
+bool Instruction::isStore() const { return isStore_; }
+bool Instruction::isLoad() const { return isLoad_; }
+bool Instruction::isBranch() const { return isBranch_; }
+bool Instruction::isRET() const { return isRET_; }
+bool Instruction::isBL() const { return isBL_; }
+bool Instruction::isSVE() const { return isSVE_; }
```

Comment:
Don't think SVE is relevant to RISC-V

Reply:
It isn't but I can't get rid of it as it is a virtual function in simeng/Instruction.hh, which has a comment about possible removal